File: packages/repo-gc-python/python/repo_gc_python/heuristics/import_diversity.py
"""Import diversity (GODF) heuristic.

Detects files that import from too many unrelated domains — a "god file"
that does too much. Backed by the Program Decomposition paper
(arXiv 2401.12412): reducing cross-file dependencies shrinks context to
~5% of window.

Logic: count distinct first-segment domains across all imported modules.
If count > limit -> finding.
Severity: >= limit*3 -> High, >= limit*2 -> Medium, >= limit -> Low.
"""

from ..cli import Threshold
from ..discovery import PythonFile
from ..parsing import FileInfo
from ..types import Finding, FindingKind, Severity


def _extract_domain(module_path: str) -> str | None:
    """Extract the top-level domain from a Python module path.

    - "os.path.join" -> "os"
    - "numpy.linalg" -> "numpy"
    - "mypackage.utils.helpers" -> "mypackage"
    - ".module" or "..parent.module" -> None (relative, skipped)
    """
    if module_path.startswith("."):
        return None  # relative import
    parts = module_path.split(".")
    return parts[0]


def analyze(
    pfile: PythonFile,
    info: FileInfo,
    threshold: Threshold,
    counter: int,
) -> Finding | None:
    limit = threshold.import_domain_limit

    domains: set[str] = set()
    for mod_path in info.imported_modules:
        domain = _extract_domain(mod_path)
        if domain:
            domains.add(domain)

    count = len(domains)

    if count <= limit:
        return None

    if count >= limit * 3:
        severity = Severity.High
    elif count >= limit * 2:
        severity = Severity.Medium
    else:
        severity = Severity.Low

    # Top 8 domains (sorted alphabetically)
    top_domains = ", ".join(sorted(domains)[:8])

    return Finding(
        id=f"id-{counter:03d}",
        kind=FindingKind.ImportDiversity,
        severity=severity,
        confidence=0.70,
        path=str(pfile.relative_path),
        summary="",
        reasons=[],
        evidence=[
            f"domain_count: {count}",
            f"domains: {top_domains}",
            f"limit: {limit}",
        ],
        suggested_next_step="",
        estimated_tokens=None,
    )

File: packages/repo-gc/src/shared/heuristics/import-diversity.ts
import { Finding, FindingKind, Severity, type FileData } from '../types';
import { type Thresholds } from '../threshold';

/**
 * Extract a "domain" from an import specifier.
 *
 * - `"./utils"` / `"../foo"` → null (relative, skipped)
 * - `"react"` → `"react"`
 * - `"lodash/merge"` → `"lodash"`
 * - `"@angular/core"` → `"@angular"`  (scoped package — first segment)
 * - `"std::collections::HashMap"` → `"std"` (Rust-style path)
 * - `"crate::foo::bar"` → `"foo"`      (skip `crate`)
 * - `"super::baz"` / `"self::qux"` → null (relative)
 */
function extractDomain(spec: string): string | null {
  if (spec.startsWith('.')) return null;

  if (spec.includes('::')) {
    const parts = spec.split('::');
    const first = parts[0];
    if (first === 'self' || first === 'super') return null;
    if (first === 'crate') {
      if (parts.length < 2) return null;
      return parts[1].split(" as ")[0].trim();
    }
    return first.split(" as ")[0].trim();
  }

  const parts = spec.split('/');
  return parts[0];
}

export function analyzeImportDiversity(
  data: FileData,
  thresholds: Thresholds,
  idCounter: { value: number },
): Finding | null {
  const limit = thresholds.importDomainLimit;
  const domains = new Set<string>();

  for (const imp of data.imports) {
    const domain = extractDomain(imp);
    if (domain) {
      domains.add(domain);
    }
  }

  const distinctDomainCount = domains.size;

  if (distinctDomainCount <= limit) {
    return null;
  }

  let severity: Severity;
  if (distinctDomainCount >= limit * 3) {
    severity = Severity.High;
  } else if (distinctDomainCount >= limit * 2) {
    severity = Severity.Medium;
  } else {
    severity = Severity.Low;
  }

  const sortedDomains = [...domains].sort();
  const topDomains = sortedDomains.slice(0, 8).join(', ');

  return {
    id: `id-${String(idCounter.value++).padStart(3, '0')}`,
    kind: FindingKind.ImportDiversity,
    severity,
    confidence: 0.70,
    path: data.relativePath,
    summary: '',
    reasons: [],
    evidence: [
      `domain_count: ${distinctDomainCount}`,
      `domains: ${topDomains}`,
      `limit: ${limit}`,
    ],
    suggested_next_step: '',
    estimated_tokens: undefined,
  };
}
